text/0021-accessibility.md

```diff
@@ -0,0 +1,485 @@
+
+```rust
+#[cfg(feature = "a11y")]
+fn a11y_nodes(&self, layout: Layout<'_>) -> iced_accessibility::A11yTree {
+    use enumset::enum_set;
+    use iced_accessibility::{
+        accesskit::{kurbo::Rect, Action, DefaultActionVerb, Node, Role},
+        A11yNode, A11yTree,
+    };
+
+    // create the child tree using the layout child
+    let child_layout = layout.children().next().unwrap();
+    let child_tree = self.content.as_widget().a11y_nodes(child_layout);
+
+    // get the bounds for the button's node
+    let Rectangle {
+        x,
+        y,
+        width,
+        height,
+    } = layout.bounds();
+    let bounds = Some(Rect::new(
+        x as f64,
+        y as f64,
+        (x + width) as f64,
+        (y + height) as f64,
+    ));
+
+    // build the node
+    // The button includes some actions. The default action is important, and many widgets may support the default action. This represents the default way of interacting with a button, in this case, clicking. For other widgets it may be different.
+    let node = Node {
+        role: Role::Button,
+        actions: enum_set!(Action::Focus | Action::Default),
+        bounds,
+        name: self.name.as_ref().map(|n| n.to_string().into_boxed_str()),
+        description: self
+            .description
+            .as_ref()
+            .map(|n| n.to_string().into_boxed_str()),
+        focusable: true,
+        default_action_verb: Some(DefaultActionVerb::Click),
+        ..Default::default()
+    };
+
+    // Create the A11yTree using the button node and its child tree using the helper
+    A11yTree::node_with_child_tree(
+        A11yNode::new(node, self.id.clone()),
+        child_tree,
+    )
+}
+```
+
+Feel free to check out the other implementations, or maybe check out the Egui implementation as well.
+
+
+## Implementation strategy
+
+
+This proposal starts with the `iced_accessibility` crate and the types that it introduces. `iced_accessibility` re-exports the accesskit crate and the chosen feature-gated platform adapter, one of `accesskit-winit` (most likely), `accesskit-unix`, `accesskit-windows`, or `accesskit-macos`. Iced-rs would likely want to maintain a fork of accesskit that sets the winit dependency of `accesskit-winit` to https://github.com/iced-rs/winit. 
+
+`iced_accessibility` also provides the following types for use in `iced_native`, `iced_winit`, `iced_glutin`, or other iced crates that need to build accessibility trees from accesskit nodes.
+
+### A11yId
+
+Accessibility Nodes need relatively consistent Ids for each element in the tree. A11yId is built using a slightly modified version of the existing `iced_native::widget::id::Id`. Because `iced_native` will depend on `iced_accessibility` the `Id` type is moved to `iced_core` and be re-exported by `iced_native`.
```

Comment:
```suggestion
Accessibility Nodes need consistent IDs for each element in the tree. A11yId is built using a slightly modified version of the existing `iced_native::widget::id::Id`. Because `iced_native` will depend on `iced_accessibility` the `Id` type is moved to `iced_core` and re-exported by `iced_native`.
```